Added telemetry that will help identify issues caused by local changes in summarizer

agarwal-navin opened the pull request:
Local changes in the summarizer (local data stores, ops) are breaking invariants in the summarizer / GC flow mainly in the summarizer node code. This is manifesting in various forms by hitting asserts (0x167, 0x1a5, 0x1b5) which makes it harder to tell what the underlying issue is.
This change converts some of these asserts into errors for 2 reasons:
1. These asserts are not valid anymore in the presence of local changes in the summarizer. There is a plan to disallow local changes in summarizer and then these asserts make sense but for now, these should not be asserts.
2. With errors, we can log additional details that can help identify the problem. For example, if these asserts are getting hit for reasons other than local changes, we want to know about them and fix them.

Changes in this PR;
1. Add id to summarizer nodes that is used for logging. This id is the same as data store / DDS id and can help correlate which summarizer node failes.
2. Changed asserts 0x1a5 and 0x1b5 to errors and logged a telemetrry. Added additional details in the log that can help identify where / what the issues are.
3. Added a log at the end of RefreshLatestSummary that tells what the result of refresh was. This can help in debugging issues where refresh was called and there are some errors later but we aren't sure if refresh is actually causing these issues.

[AB#2375](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/2375), [AB#2767](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/2767), [AB#3392](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/3392)

agarwal-navin committed:
Added telemetry that whelp identify issues caused by local changes in summarizer

agarwal-navin commented on packages/runtime/runtime-utils/src/summarizerNode/summarizerNode.ts:
```diff
@@ -61,10 +62,40 @@ export class SummarizerNode implements IRootSummarizerNode {
+		private readonly initialSummary?: IInitialSummary,
+		protected wipSummaryLogger?: ITelemetryLogger,
+		/** A unique id of this node to be logged when sending telemetry. */
+		protected telemetryId?: string,
```
Moved the constuctor to the top because I can never find it. The only change here is addition of `telemetryId` and the child logger creation in the constructor.

vladsud replied:
should it be called nodeId? or something else that more clearly indicates what this ID is about?

agarwal-navin replied:
Renamed it to "telemetryNodeId". I added the prefix telemetry because it should not be used for anything other than logging.

agarwal-navin commented on packages/runtime/runtime-utils/src/summarizerNode/summarizerNode.ts:
```diff
@@ -235,66 +288,100 @@ export class SummarizerNode implements IRootSummarizerNode {
+				wasSummaryTracked?: boolean;
+				summaryRefSeq?: number;
+			} = { ...result };
+			this.logger.sendTelemetryEvent({
```
@tyler-cai-microsoft Added the log that will tell the result of refresh latest summary.

agarwal-navin commented on packages/runtime/runtime-utils/src/summarizerNode/summarizerNode.ts:
```diff
@@ -235,66 +288,100 @@ export class SummarizerNode implements IRootSummarizerNode {
-		if (proposalHandle !== undefined) {
-			const maybeSummaryNode = this.pendingSummaries.get(proposalHandle);
+		let result: RefreshSummaryResult | undefined;
+		try {
```
The change here is the addition of try ... finally block to add the "refreshLatestSummary_end" log at the end. There is no change in the logic.

tyler-cai-microsoft commented on packages/runtime/runtime-utils/src/summarizerNode/summarizerNode.ts:
```diff
@@ -235,66 +288,100 @@ export class SummarizerNode implements IRootSummarizerNode {
+		 * Refresh latest summary should never happen while a summary is in progress. If it does, log an error
+		 * so that we can gather data around it and take appropriate action.
+		 */
+		if (this.isTrackingInProgress()) {
```
Is this a specific GC summarizer node thing? Also, maybe valuable to converge both the summarizerNode and the summarizerNodeWithGC together.

agarwal-navin replied:
isTrackingInProgress() is a generic summarizer node API.
No, we intentionally kept summarizer node and summarizerNodeWithGC separate since summarizer node is a generic base imlement that tracks summaries. You can build any number of state tracking on top of it including GC.

vladsud commented on packages/runtime/container-runtime/src/dataStores.ts:
```diff
@@ -563,11 +564,26 @@ export class DataStores implements IDisposable {
-					);
+					// Summarizer works only with clients with no local changes. A data store in attaching
+					// state indicates an op was sent to attach a local data store.
+					if (context.attachState === AttachState.Attaching) {
```
Can you please double check that we get here only in summarizer client? I.e. it's not hit in various container serialization or on-demand summary cases.

vladsud commented on packages/runtime/runtime-utils/src/summarizerNode/summarizerNode.ts:
```diff
@@ -176,10 +207,32 @@ export class SummarizerNode implements IRootSummarizerNode {
+		 * will help us identify these cases and take appropriate action.
+		 */
+		if (localPathsToUse === undefined) {
+			const error = new LoggingError("NodeNotSummarized", {
```
same question as before - can we somehow reuse code and reduce impact on bundle size?

vladsud commented on packages/runtime/runtime-utils/src/summarizerNode/summarizerNode.ts:
```diff
@@ -235,66 +288,100 @@ export class SummarizerNode implements IRootSummarizerNode {
-				eventName: "PendingSummaryNotFound",
+			return result;
+		} finally {
+			const loggingProps: {
```
Should you use PerformanceEvent here instead of manually controlling start / end / exception flow?

agarwal-navin replied:
Will do. I didn't know that I could pass a param to performance event to also log the start event.

vladsud commented:
Actually, can we keep asserts on summarization path, and instead throw user-oriented errors much earlier in pipeline? I.e. when local DDS / datastore are created? That way we would properly be throwing UsageError, would not need to track IDs (as we do not have them yet), and stack would be much more actionable.

agarwal-navin commented:
> Actually, can we keep asserts on summarization path, and instead throw user-oriented errors much earlier in pipeline? I.e. when local DDS / datastore are created? That way we would properly be throwing UsageError, would not need to track IDs (as we do not have them yet), and stack would be much more actionable.

Yes, that's the goal. I already added logging when local changes are done in the summarizer. There will be a follow up that throws errors during local changes as well. Here's why changing them to errors is useful for now:
As seen in telemetry, these asserts are being hit. I am pretty sure its because of local data stores but I can't say that with 100% certainty without the logging that I added in https://github.com/microsoft/FluidFramework/pull/14070. So, until we can find and fix all issues that lead to these asserts, error gives us more data in telemetry. We will convert them back to asserts later.

agarwal-navin committed:
PR comments integration

agarwal-navin commented:
Actually, I changed the errors in summarize() and getGCData() to asserts since it won't give us any additional information. We know for sure this happens because of local data stores and we have logging that will show it.
The summarizer node ones are still errors just to confirm that the error are indeed because of local data stores only.

agarwal-navin committed:
Remove redundant log

agarwal-navin commented on packages/runtime/runtime-utils/src/summarizerNode/summarizerNode.ts:
```diff
@@ -235,66 +282,104 @@ export class SummarizerNode implements IRootSummarizerNode {
+				// Refresh latest summary should not happen while a summary is in progress. If it does, it can result
+				// in inconsistent state, so, we should not continue;
+				if (this.isSummaryInProgress()) {
+					throw new LoggingError("UnexpectedRefreshDuringSummarize", {
```
@NicholasCouri I added this validation here. Do you think I should wait until https://dev.azure.com/fluidframework/internal/_workitems/edit/3385 is fixed?
